crates/spawner/src/destroyer.rs

```diff
@@ -1,33 +1,253 @@
+        app.add_event::<DespawnEvent>()
+            .add_system(
+                find_dead
+                    .in_base_set(GameSet::Update)
+                    .run_if(in_state(AppState::InGame))
+                    .before(DespawnerSet::Despawn),
+            )
+            .add_system(
+                despawn
+                    .in_base_set(GameSet::Update)
+                    .run_if(in_state(AppState::InGame))
+                    .in_set(DespawnerSet::Despawn),
+            );
     }
 }
 
-fn destroy(
-    mut commands: Commands,
+#[derive(Debug, Clone, PartialEq, Eq, Hash, SystemSet)]
+pub enum DespawnerSet {
+    /// This set is run before the despawning systems run (If you expect units to die, you should
+    /// run your system in this set and the `Update` base set)
+    Destruction,
+    Despawn,
+}
+
+struct DespawnEvent(Entity);
+
+// Find all entities with low health and mark them for despawning
+fn find_dead(
     mut counter: ResMut<ObjectCounter>,
     entities: Query<(Entity, &Player, &ObjectType, &Health), Changed<Health>>,
+    mut event_writer: EventWriter<DespawnEvent>,
 ) {
     for (entity, &player, &object_type, health) in entities.iter() {
         if health.destroyed() {
             if let ObjectType::Active(active_type) = object_type {
                 counter.player_mut(player).unwrap().update(active_type, -1);
             }
-            commands.entity(entity).despawn_recursive();
+            event_writer.send(DespawnEvent(entity));
+        }
+    }
+}
+
+fn despawn(mut commands: Commands, mut despawning: EventReader<DespawnEvent>) {
+    for entity in despawning.iter() {
+        commands.entity(entity.0).despawn_recursive();
+    }
+}
+
+#[derive(Debug)]
+pub struct DespawnEventsPlugin<Q, T, F = ()>
+where
+    F: ReadOnlyWorldQuery + Send + Sync + 'static,
+    T: Send + Sync + 'static + Clone + Component,
+    Q: WorldQuery + Send + Sync + 'static,
+{
+    _marker: PhantomData<(Q, F, T)>,
+}
+
+impl<
+        Q: WorldQuery + Send + Sync + 'static,
+        T: Send + Sync + 'static + Clone + Component,
+        F: ReadOnlyWorldQuery + Send + Sync + 'static,
+    > Default for DespawnEventsPlugin<Q, T, F>
+{
+    fn default() -> Self {
+        Self {
+            _marker: PhantomData,
+        }
+    }
+}
+
+impl<
+        Q: WorldQuery + Send + Sync + 'static,
+        T: Send + Sync + Clone + Component + 'static,
+        F: ReadOnlyWorldQuery + Send + Sync + 'static,
+    > Plugin for DespawnEventsPlugin<Q, T, F>
+{
+    fn build(&self, app: &mut App) {
+        app.add_event::<DespawnedComponents<T, F>>().add_system(
+            send_data::<Q, T, F>
+                .in_base_set(GameSet::Update)
+                .run_if(in_state(AppState::InGame))
+                .after(DespawnerSet::Destruction)
+                .in_set(DespawnerSet::Despawn),
+        );
+    }
+}
+
+// Contains the entity and the data of the despawned component
+#[derive(Debug)]
+pub struct DespawnedComponents<T, F = ()>
```

Comment:
I would rename this to `DespawnedComponentsEvent` so it is in accordance with the style guide (see CONTRIBUTING.md).